uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-assignment.controller.js

```diff
@@ -259,6 +261,91 @@ define([
       });
     }
 
+    /**
+     * Returns the default assignee for the activity type provided.
+     *
+     * @param {Object} activityType the activity type definition.
+     * @return {Promise} resolves to an array wrapping the default assignee's contact id or NULL in case of none.
+     */
+    function getDefaultAssigneeForActivityType (activityType) {
+      switch (activityType.default_assignee_type) {
+        case defaultAssigneeOptionsIndex.SPECIFIC_CONTACT:
+          return $q.resolve([ activityType.default_assignee_contact ]);
+        case defaultAssigneeOptionsIndex.USER_CREATING_THE_CASE:
+          return $q.resolve([ session.contactId ]);
+        case defaultAssigneeOptionsIndex.BY_RELATIONSHIP:
+          return getDefaultAssigneeForActivityTypeByRelationship(activityType);
+        case defaultAssigneeOptionsIndex.NONE:
+        default:
+          return $q.resolve(null);
+      }
+    }
+
+    /**
+     * Returns the default assignee depending on the relationship to the selected
+     * target contact.
+     *
+     * @param {Object} activityType the activity type definition.
+     * @return {Promise}
+     */
+    function getDefaultAssigneeForActivityTypeByRelationship (activityType) {
+      // skip if a target contact has not been selected:
+      if (!$scope.assignment.client_id) {
+        return $q.resolve(null);
+      }
+
+      return Relationship.allValid({
+        'contact_id_a': $scope.assignment.client_id,
+        'relationship_type_id.name_b_a': activityType.default_assignee_relationship,
+        'options': { 'limit': 1 }
+      })
+        .then(function (result) {
+          return result.list.map(function (relationship) {
+            return relationship.contact_id_b;
+          });
+        });
+    }
+
+    /**
+     * Initializes the default assignee for each one of the tasks available.
+     */
+    function initDefaultAssigneesForTasks () {
+      // skip if an activity set and types have not been defined:
+      if (!$scope.activity.activitySet.activityTypes) {
+        return;
+      }
+
+      $scope.activity.activitySet.activityTypes.forEach(function (activityType) {
+        var task = _.find($scope.taskList, function (task) {
+          return task.name === activityType.name;
+        });
+
+        if (!task) {
+          return;
+        }
+
+        getDefaultAssigneeForActivityType(activityType)
+          .then(function (assigneeId) {
+            task.assignee_contact_id = assigneeId;
+          })
+          .then(function () {
+            return loadAndCacheContactForTask(task);
```

Comment:
Instead of loading each contact name in a separate request, we need to do it in a single time. Here is the algo
1. From `$scope.taskList`, prepare a list of all contacts, and get the contact names using `IN` parameter of API. And cache it in a object. Do it before calling `initDefaultAssigneesForTasks`

2. Get the names synchronously using the object saved in previous state.

Reply:
Agreed, though it's a bit more complicated than that since the cache is for each individual task, so after I get the contacts from that single request I'd have to go task by task and match them to the user result. I'll work on it.